src/app/clusters/ota-requestor/OTARequestor.cpp

```diff
@@ -84,84 +55,138 @@ OTARequestorInterface * GetRequestorInstance()
-}
-// Finds the Requestor instance and calls the corresponding OTARequestor member function
-void OnQueryImageResponse(void * context, const QueryImageResponse::DecodableType & response)
-{
-    OTARequestor * requestorCore = static_cast<OTARequestor *>(GetRequestorInstance());
-
-    assert(requestorCore != nullptr);
+    // Check against minimum length of a valid BDX URI
+    if (uri.size() < kValidBdxUriMinLen)
+    {
+        ChipLogError(SoftwareUpdate, "Expect BDX URI to be at least %d in length, actual length is %zu", kValidBdxUriMinLen,
+                     uri.size());
+        return CHIP_ERROR_INVALID_STRING_LENGTH;
+    }
 
-    requestorCore->mOnQueryImageResponse(context, response);
-}
+    // Check the scheme field matches the BDX prefix
+    CharSpan expectedScheme(bdxPrefix, strlen(bdxPrefix));
+    if (!uri.SubSpan(0, strlen(bdxPrefix)).data_equal(expectedScheme))
+    {
+        ChipLogError(SoftwareUpdate, "Expect scheme to contain prefix %s, actual URI: %.*s", bdxPrefix,
+                     static_cast<int>(uri.size()), uri.data());
+        return CHIP_ERROR_INVALID_SCHEME_PREFIX;
+    }
 
-void OTARequestor::mOnQueryImageResponse(void * context, const QueryImageResponse::DecodableType & response)
-{
-    ChipLogDetail(SoftwareUpdate, "QueryImageResponse responded with action %" PRIu8, response.status);
+    // Extract the node ID from the authority field
+    CharSpan nodeIdString = uri.SubSpan(strlen(bdxPrefix), kNodeIdHexStringMaxLen);
+    uint8_t buffer[kNodeIdHexStringMaxLen];
+    if (chip::Encoding::HexToBytes(nodeIdString.data(), nodeIdString.size(), buffer, kNodeIdHexStringMaxLen) == 0)
+    {
+        ChipLogError(SoftwareUpdate, "Cannot convert image URI provider node ID: %.*s", static_cast<int>(nodeIdString.size()),
+                     nodeIdString.data());
+        return CHIP_ERROR_INVALID_DESTINATION_NODE_ID;
+    }
 
-    CHIP_ERROR err = CHIP_NO_ERROR;
+    nodeId = chip::Encoding::BigEndian::Get64(buffer);
+    if (!IsOperationalNodeId(nodeId))
+    {
+        ChipLogError(SoftwareUpdate, "Image URI contains invalid node ID: 0x" ChipLogFormatX64, ChipLogValueX64(nodeId));
+        return CHIP_ERROR_INVALID_DESTINATION_NODE_ID;
+    }
 
-    // TODO: handle QueryImageResponse status types
-    if (response.status != EMBER_ZCL_OTA_QUERY_STATUS_UPDATE_AVAILABLE)
+    // Verify the separator between authority and path fields
+    CharSpan separator(bdxSeparator, strlen(bdxSeparator));
+    if (!uri.SubSpan(strlen(bdxPrefix) + kNodeIdHexStringMaxLen, strlen(bdxSeparator)).data_equal(separator))
     {
-        return;
+        return CHIP_ERROR_MISSING_URI_SEPARATOR;
     }
 
-    VerifyOrReturn(mBdxDownloader != nullptr, ChipLogError(SoftwareUpdate, "downloader is null"));
+    // Extract file designator from the path field
+    size_t fileDesignatorLength = uri.size() - (strlen(bdxPrefix) + kNodeIdHexStringMaxLen + strlen(bdxSeparator));
```

Comment:
A lot of this arithmetic would be quite a bit easier with a `Encoding::LittleEndian::Reader` against the URI buffer, and repeated calls to `ReadBytes()` and checking for `StatusCode()`. This would avoid needing to do any arithmetic as the cursor management would be done by teh reader.

Reply:
This entire function has been moved but I did switch to using Reader now